kpatch-build/create-diff-object.c

```diff
@@ -1213,8 +1232,39 @@ static void kpatch_replace_sections_syms(struct kpatch_elf *kelf)
 
 				rela->sym = sym;
 				rela->addend -= start;
+				found = 1;
 				break;
 			}
+			/*
+			 * Unbundled sections might refer to SHN_UNDEF symbol type, example kernel/sched/core.o.
+			 * Relocation section '.rela.data' at offset 0xadda8 contains 2 entries:
+			 *    Offset             Info             Type       Symbol's Value   Symbol's Name + Addend
+			 * 0000000000000000  000003ae00000026 R_PPC64_ADDR64 0000000000000000 call_rcu_sched + 0
+			 * 0000000000000008  000003ae00000026 R_PPC64_ADDR64 0000000000000000 call_rcu_sched + 0
```

Comment:
In the example stated above, the `.toc section` of `sched/core.o` there is a relocation entry `.data + 0`. Where the attempt is made to replace the symbol + offset with appropriate symbol from `.data section`, this attempt fails:
- `.data` is a section and symbol `call_rcu_sched` does not belong to any section.
-  `call_rcu_sched` is of  relocation type `R_PPC64_ADDR64`  and bind type `SHN_UNDEF`. Which is a global symbol.
The replacement for such symbols need extra step to be replaced with the `section + offset` directly.

Reply:
Ok.  Can you also show the .toc section `.data + 0` entry in the comment?

Reply:
This comment is still confusing.  It should show what is being replaced with what.

Also, if I understand correctly, it's replacing the .toc reference with a symbol reference?  If so, why is that needed?  I thought we were leaving such relocations alone.

Reply:
>Also, if I understand correctly, it's replacing the .toc reference with a symbol reference? If so, why is that needed? I thought we were leaving such relocations alone.

The `kpatch_replace_sections_syms` attempt to replace `.toc section + 0x468`  section symbol, fails:
`0000000000000468  0000011d00000026 R_PPC64_ADDR64         0000000000000000 .data + 0`

Reason being the symbol at relocation entry in `.data` section a.k.a `call_rcu_sched` is of type `SHN_UNDEF`/bind type `STB_GLOBAL`.  It doesn't belong to `.data` section or any section in the object file.  When `kpatch_include_standard_elements` includes every relocation symbols of the `.toc section` by calling `kpatch_include_symbol`. Inclusion of `.data` symbol, also includes the `.data` section.  Attempt to include `.data` section fails with:
```
core.o: changed function: sched_cpu_deactivate
core.o: data section .data selected for inclusion
ERROR: core.o: 1 unsupported section change(s)
```
Does the below replacement comment explains the explain better, than the one in the patch
```
/*
 * Relocation section '.rela.toc' at offset 0xad9f8 contains 162 entries:
 *    Offset             Info             Type       Symbol's Value    Symbol's Name + Addend
 * ...
 * 0000000000000468  0000011d00000026 R_PPC64_ADDR64 0000000000000000  .data + 0
 * 
 * Relocation section '.rela.data' at offset 0xadda8 contains 2 entries:
 *    Offset             Info             Type       Symbol's Value   Symbol's Name + Addend
 * 0000000000000000  000003ae00000026 R_PPC64_ADDR64 0000000000000000 call_rcu_sched + 0
 *
 * Replacement of unbundled section with global symbols, needs extra work of iterating through
 * the relocation entries of the rela->sym->sec. The reason being such sym->sec is always NULL.
 */
```

Reply:
I don't think this is the right approach.  `.data + 0` is a data pointer.  If you look at the code, it's trying to get a pointer to *a pointer to* a function.  See the _wait_rcu_gp() macro.  It creates a __crcu_array[], which is an array of function pointers.

But if I understand correctly, you're converting the `__crcu_array[]` reference to `__crcu_array[0]`, which is dangerous.

Reply:
Reference to `.data + 0` is the first element of the array of pointers, consider the below: (compiled with -O1 for readability). At the offset 0x80, the load of the second element/function pointer is `data[0] + 0x8` i.e first element + function pointer size. In theory, this might work any suggestions on how do we make it safer.
```                        
  74:   00 00 42 3d     addis   r10,r2,0
                        74: R_PPC64_TOC16_HA    .toc+0x458
  78:   00 00 2a e9     ld      r9,0(r10)
                        78: R_PPC64_TOC16_LO_DS .toc+0x458
  7c:   00 00 49 e9     ld      r10,0(r9)
  80:   08 00 69 e9     ld      r11,8(r9)
  84:   80 00 41 f9     std     r10,128(r1)
  88:   88 00 61 f9     std     r11,136(r1)
  8c:   20 00 c1 38     addi    r6,r1,32
  90:   80 00 a1 38     addi    r5,r1,128
  94:   02 00 80 38     li      r4,2
  98:   00 00 60 38     li      r3,0
  9c:   01 00 00 48     bl      9c <sched_cpu_deactivate+0x94>
                        9c: R_PPC64_REL24       __wait_rcu_gp
  a0:   00 00 00 60     nop
```

Reply:
> In theory, this might work

But how would that work?  With this code, r9 would have the address of `call_rcu_sched` and then
```
  7c:   00 00 49 e9     ld      r10,0(r9)
  80:   08 00 69 e9     ld      r11,8(r9)
```
would load memory from the function text, right?

Reply:
Thanks for being patient with me on this. True, `.data` being an array of function pointers and the relocation entries of `.rela.data` section are of `R_PPC64_ADDR64` type. One idea is to include `.data` section, which will get included as part of including the section of the symbols in `.toc` section. While checking for the verifiability of the patch, a check can be added to check if the relocation entries are function pointers only.
```
--- a/kpatch-build/create-diff-object.c
+++ b/kpatch-build/create-diff-object.c
@@ -373,6 +373,19 @@ static int rela_equal(struct rela *rela1, struct rela *rela2)
        return !strcmp(rela1->sym->name, rela2->sym->name);
 }
 
+static int has_only_func_ptr(struct section *sec)
+{
+       struct rela *rela;
+
+       if (!sec->rela)
+               return 0;
+
+       list_for_each_entry(rela, &sec->rela->relas, list)
+               if (rela->sym->sym.st_shndx != SHN_UNDEF)
+                       return 0;
+       return 1;
+}
+
 static void kpatch_compare_correlated_rela_section(struct section *sec)
 {
        struct rela *rela1, *rela2 = NULL;
@@ -1305,7 +1319,8 @@ static void kpatch_verify_patchability(struct kpatch_elf *kelf)
                 * (.data.unlikely and .data.once is ok b/c it only has __warned vars)
                 */
                if (sec->include && sec->status != NEW &&
-                   (!strncmp(sec->name, ".data", 5)     || !strncmp(sec->name, ".bss", 4)) &&
+                   ((!strncmp(sec->name, ".data", 5) && !has_only_func_ptr(sec)) ||
+                    !strncmp(sec->name, ".bss", 4)) &&
                    (strcmp(sec->name, ".data.unlikely") && strcmp(sec->name, ".data.once"))) {
                        log_normal("data section %s selected for inclusion\n",
                                   sec->name);
```

Reply:
We can't to include the .data section, because we don't support changes to data.  Instead we need to replace the .data relocation with a dynrela, so that it refers to the original array.

Reply:
Agree, right approach should be to generate a dynrela replacing the reference with `STT_OBJECT` symbol which is generally the name of the array and here it should have been `__crcu_array`. This seems to be an exception to this rule, where gcc doesn't generate a symbol with `STT_OBJECT` type. Whereas load of the first element in the array is done via `.data sec + offset`. In such case, it not possible to refer to the original array. Please correct me, if I am wrong here.

Reply:
> This seems to be an exception to this rule, where gcc doesn't generate a symbol with STT_OBJECT type. Whereas load of the first element in the array is done via .data sec + offset. In such case, it not possible to refer to the original array. Please correct me, if I am wrong here.

Does kpatch_replace_sections_syms() not convert such a reference to an object reference?

Reply:
> Does kpatch_replace_sections_syms() not convert such a reference to an object reference?

kpatch_replace_sections_syms() doesn't convert the reference of such symbols, those do not belong to any section a.k.a `SHN_UNDEF`. I am waiting for thoughts from toolchains developers on how to resolve such data variable. Can we track, this particular patch in a new issue and merge other patches?

Reply:
Sure, feel free to split up the patches however you like.  Also I don't think I understand this issue very well, maybe attaching the .o file would help.

Reply:
> Sure, feel free to split up the patches however you like. Also I don't think I understand this issue very well, maybe attaching the .o file would help.

Thank you. Will drop this patch and request the pull request again. Will open a PR for this issue, with the object file.